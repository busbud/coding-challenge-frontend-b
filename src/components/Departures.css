.departures {
  max-width: 980px;
  margin: 0 auto;
  padding: 0px 40px; }
  .departures .result-number {
    margin-top: 20px; }
  .departures .departure {
    margin: 1.42rem 0;
    padding: 24px 38px;
    border: none;
    box-shadow: 0 1px 3px 0 rgba(0, 0, 0, 0.24);
    border-radius: 4px;
    background: white;
    display: flex;
    justify-content: space-between;
    align-items: center; }
    .departures .departure .departure-summary span.time {
      margin-right: 15px;
      font-weight: bold; }
    .departures .departure .departure-summary span.city strong:after {
      content: ' - '; }
    .departures .departure .departure-summary .departure-summary-origin span.time {
      color: #127ccb; }
    .departures .departure .departure-summary .departure-summary-destination {
      margin-top: 10px;
      color: rgba(32, 37, 43, 0.35); }
    .departures .departure .departure-prices {
      color: #f19020;
      font-weight: 700;
      font-size: 24px; }
      .departures .departure .departure-prices small {
        color: rgba(241, 144, 32, 0.54);
        font-weight: 400; }

@media (max-width: 600px) {
  .departures {
    padding: 0px 20px;
    margin-top: 1.42rem; }
    .departures .departure {
      margin: 0px;
      padding: 20px;
      border-radius: 0px; } }
