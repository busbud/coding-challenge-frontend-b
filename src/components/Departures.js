import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createStructuredSelector } from 'reselect';
// import InfiniteScroll from 'react-infinite-scroll-component';

import InfiniteScroll from 'react-infinite-scroller';

import Departure from './Departure';
import Loader from './Loader';

import {
  initDepartures, pollDepartures, departuresSelector, isCompleteSelector, isLoadingSelector,
} from '../store/modules/departures';

const Departures = ({
  departures, isComplete, isLoading, initDeparturesConnect, pollDeparturesConnect,
}) => {
  useEffect(() => {
    initDeparturesConnect();
  }, []);

  const pollData = () => {
    if (!isComplete && !isLoading) {
      pollDeparturesConnect();
    }
  };

  return (
    <div className="">

      <InfiniteScroll
        pageStart={0}
        loadMore={pollData}
        hasMore={!isComplete}
        // initialLoad={false}
        loader={(
          <div style={{ textAlign: 'center', clear: 'both' }}>
            <Loader />
          </div>
        )}
        useWindow={false}
      >
        {departures.map(departure => (
          <Departure departure={departure} />
        ))}
      </InfiniteScroll>
      {isComplete && !isLoading && (
        <p style={{ textAlign: 'center' }}>
          <b>No more available departures</b>
        </p>
      )}
    </div>
  );
};

Departures.propTypes = {
  departures: PropTypes.array,
  isComplete: PropTypes.bool.isRequired,
  isLoading: PropTypes.bool.isRequired,
  initDeparturesConnect: PropTypes.func.isRequired,
  pollDeparturesConnect: PropTypes.func.isRequired,
};

Departures.defaultProps = {
  departures: [],
};

const mapStateToPropsSelector = createStructuredSelector({
  departures: departuresSelector,
  isComplete: isCompleteSelector,
  isLoading: isLoadingSelector,
});

function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    initDeparturesConnect: initDepartures,
    pollDeparturesConnect: pollDepartures,
  }, dispatch);
}

export default connect(mapStateToPropsSelector, mapDispatchToProps)(Departures);
