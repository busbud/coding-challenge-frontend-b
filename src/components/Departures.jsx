import React from 'react';
import Loading from './Loading';

export default class Departures extends React.Component {
  render() {
    const { schedules } = this.props
    return (
      <div className='depatures'>
        { schedules.departures.map( (depature, index) => <Depature key={index} depature={depature} /> ) }
        { schedules.isLoading && <Loading /> }
      </div>
    )
  }
}

const Depature = ({depature}) => (
  <div className='depature'>
    {depature.id}
  </div>
)
