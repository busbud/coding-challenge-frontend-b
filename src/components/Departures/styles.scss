@import '../../shared-styles/variables';

.subheading {
  font-size: 2 * $padding;
  line-height: 2.5 * $padding;
  text-align: center;
  margin-bottom: 2.5 * $padding;
}

.data {
  font-weight: bold;
}
