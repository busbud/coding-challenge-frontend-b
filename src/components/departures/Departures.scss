.Departures {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
    justify-content: center;
    background-color: #F6F6F6;

    p, div {
        font-family: Arial, Helvetica, sans-serif
    }
}