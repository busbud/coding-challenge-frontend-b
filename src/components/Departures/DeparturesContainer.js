import React, { Component } from 'react';
import arrayFind from 'array-find';
import objectAssign from 'object-assign';

import Departures from './Departures';
import fetchDepartures from '../../services/clientFetchDepartures';

class DeparturesContainer extends Component {
  constructor(props) {
    super(props);
    /* Most of this information is static
     * but it's the sort of information that
     * could start changing, so storing it as
     * state is good.
     */
    this.state = {
      origin: {
        name: 'New York',
        geohash: 'dr5reg',
      },
      destination: {
        name: 'MontrÃ©al',
        geohash: 'f25dvk',
      },
      date: '2 July 2017',
      language: 'en',
      currency: 'CAD',
      adultCount: 1,
      departures: [],
      complete: false,
    };
  }

  componentDidMount() {
    this.fetch();
  }

  fetch(index) {
    const { origin, destination, date, language, currency, adultCount } = this.state;
    fetchDepartures({
      origin: origin.geohash,
      destination: destination.geohash,
      outboundDate: new Date(date).toISOString().slice(0, 10),
      adult: adultCount,
      lang: language,
      currency,
      index,
    }).then(data => {
      this.setState({
        departures: [
          ...this.state.departures,
          ...data.departures.map(departure => {
            const operator = arrayFind(data.operators, operator =>
              departure.operator_id === operator.id
            );
            return objectAssign({}, departure, {
              operator,
            });
          }),
        ],
        complete: data.complete,
      }, () => {
        if (!this.state.complete) {
          this.fetch(this.state.departures.length, 200);
        }
      });
    }).catch(err => console.error(err)) // TODO: real error handling
  }

  render() {
    return <Departures {...this.state} />;
  }
}

export default DeparturesContainer;
