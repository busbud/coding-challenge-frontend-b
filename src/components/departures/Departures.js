import React, { Fragment } from 'react';
import DepartureDetails from '../departuresDetails/DeparturesDetails';
import { connect } from 'react-redux';

import './Departures.scss';

export const Departures = (props) => {
    return (
        <div className="Departures">
            {props.departures.map((departure, index) => {
                return (
                    <DepartureDetails
                        key={index}
                        departure={departure}
                        originLocation={props.locations[departure.origin_location_id]}
                        destinationLocation={props.locations[departure.destination_location_id]}
                    />
                )
            })}
        </div>
    )
}

const mapStateToProps = state => ({
    departures: state.departures,
    locations: state.locations
});
export default connect(mapStateToProps)(Departures);