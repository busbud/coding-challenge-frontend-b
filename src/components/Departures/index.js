import React from 'react';
import DeparturesContainer from './DeparturesContainer';

export default () => <DeparturesContainer />;
