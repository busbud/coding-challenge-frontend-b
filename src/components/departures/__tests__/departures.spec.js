import React from 'react';
import { Departures } from '../Departures';
import renderer from 'react-test-renderer';
jest.mock('../../departuresDetails/DeparturesDetails', () => 'DeparturesDetails');

const props = {
    departures: [{
        origin_location_id: 221,
        destination_location_id: 223
    }],
    locations: {
        '221': {
            name: "gare1"
        },
        '223': { name: "gare2" }
    }
}

describe('Departures', () => {
    it('renders correctly', () => {
        const tree = renderer
            .create(<Departures {...props} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
})
