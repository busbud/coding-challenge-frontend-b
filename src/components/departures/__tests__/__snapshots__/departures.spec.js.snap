// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Departures renders correctly 1`] = `
<div
  className="Departures"
>
  <DeparturesDetails
    arrival_city="MontrÃ©al"
    departure={
      Object {
        "destination_location_id": 223,
        "origin_location_id": 221,
      }
    }
    departure_city="New York"
    destinationLocation={
      Object {
        "name": "gare2",
      }
    }
    originLocation={
      Object {
        "name": "gare1",
      }
    }
  />
</div>
`;
